### ISSUE TITLE
Emit nullable annotations for c# 8.0

### ISSUE THREAD

**Issue by YairHalberstadt** [antlr/antlr4#2639]
Currently antlr generates nullable attributes to indicate nullability when targeting C#.

It would be extremely useful if there was an option to emit C# 8.0 nullable annotations instead.

**Comment by YairHalberstadt** [comment]
@sharwell 

I wonder if you have any thoughts on how difficult this would be to cover completely? I imagine the first step would be to cover the nuget packages with nullable annotations - I'm happy to do that, but I don't know if I have enough time to add an option to emit nullable annotations.

**Comment by sharwell** [comment]
@YairHalberstadt I'm planning to update tunnelvisionlabs/antlr4cs using the nullability infrastructure we created for tunnelvisionlabs/antlr4ts.

**Comment by KvanTTT** [comment]
Does the issue https://github.com/antlr/antlr4/issues/1972 influence this one? There is no distinction between missing token due to the optional operator (`?`) and due to a syntax error.

**Comment by sharwell** [comment]
@KvanTTT That issue impacts the usability of the syntax tree when errors occur.

**Comment by yonigibbs** [comment]
@sharwell, can I please just double-check the latest state of this issue? Am I right in saying that currently generated C# code will not distinguish between a value which might be null and which might not (according to the grammar schema)? i.e. the generated C# will always say something like `public SomeContext someThing;` and never `public SomeContext? someThing;`. That's what I'm seeing on my system, but not sure if I'm just doing something wrong. Thanks!

**Comment by sharwell** [comment]
> Am I right in saying that currently generated C# code will not distinguish between a value which might be null and which might not (according to the grammar schema)?

Yes, that's right.

**Comment by yonigibbs** [comment]
Cool, thanks!

**Comment by KvanTTT** [comment]
If nullable annotations are implemented maybe it will make sense to use the newest C# version for other constructions or at least C# 8.0? For instance, expression bodies for single-line properties.

**Comment by geekley** [comment]
Yes please! This is a necessity!
Unfortunately C# is not very pedantic with warnings when using unclear nullability methods in a file with nullable enabled. I wish there was a pedantic mode that would treat unclear nullability as nullable, so I could explicitly check null or assert non-null myself.
Since it's not like this, it's very confusing to have mixed nullability, cause I never know when I have a guarantee or not.

**Comment by NickStrupat** [comment]
> Yes please! This is a necessity!
> Unfortunately C# is not very pedantic with warnings when using unclear nullability methods in a file with nullable enabled. I wish there was a pedantic mode that would treat unclear nullability as nullable, so I'd have to explicitly check null or assert non-null myself.
> Since it's not like this, it's very confusing to have mixed nullability, cause I never know when I have a guarantee or not.

I totally agree. The nullable reference types in C# are basically a `preview` feature that got released. So many things broken with them.
